packages/editor/src/components/post-trash/index.js

```diff
@@ -2,19 +2,32 @@
  * WordPress dependencies
  */
 import { __ } from '@wordpress/i18n';
+import { useState } from '@wordpress/element';
 import { Button } from '@wordpress/components';
 import { withSelect, withDispatch } from '@wordpress/data';
 import { compose } from '@wordpress/compose';
 
 function PostTrash( { isNew, postId, postType, ...props } ) {
+	const [ disabled, setDisabled ] = useState( false );
+
 	if ( isNew || ! postId ) {
 		return null;
 	}
+	const onClick = () => {
+		if ( disabled ) {
+			return false;
+		}
+
+		setDisabled( true );
 
-	const onClick = () => props.trashPost( postId, postType );
+		return props.trashPost( postId, postType )
+			.then( () => {
+				setDisabled( false );
```

Comment:
Are we concerned about this being called after component unmount? For example, clicking the button and closing the sidebar right away. Won't this trigger a React warning?

Reply:
We should be concerned, I think, even if in all likelihood the trash button isn't going anywhere in the editor. There's still nothing which would stop this `setState` from happening after a component is unmounted.

For me, ideally we don't use this as a promise, and we have some state which indicates whether a "Trash" is in-progress. That way, rather than explicitly setting the local component's state, we can assign `aria-disabled` based on whether the post is currently being moved to trash.

Reply:
Based on the current implementation of how `trashPost` works, I don't think there's really any immediately-available method to determine if the current post is currently being moved to trash.

https://github.com/WordPress/gutenberg/blob/a60554d1369f8a88202d937eea2434db8308d3fe/packages/editor/src/store/actions.js#L315-L337

For me, ideally there would be one of...

1. Some state value which holds whether a save is in progress (similar to [`isSavingMetaBoxes`](https://github.com/WordPress/gutenberg/blob/a60554d1369f8a88202d937eea2434db8308d3fe/packages/edit-post/src/store/reducer.js#L193-L212))
2. The current post value is immediately updated so that it's local status is marked as `status: 'trash'`
   - Might be possible by calling `yield editPost( { status: 'trash' } )` somewhere early in this function ([action reference](https://github.com/WordPress/gutenberg/blob/a60554d1369f8a88202d937eea2434db8308d3fe/packages/editor/src/store/actions.js#L184-L204))
3. Add entity trashing as a feature of the [Core Data module](https://github.com/WordPress/gutenberg/blob/master/packages/core-data/README.md), then detect `isTrashing` by some selector.

The second of these could be promising. I like the idea that the internal data of the post is updated to reflect the trash status as soon as the button is clicked.

Otherwise, I think we have to go down the road of [this sort of cancelable promises behavior](https://github.com/WordPress/gutenberg/blob/a60554d1369f8a88202d937eea2434db8308d3fe/packages/block-editor/src/components/link-control/index.js#L55-L80), which we should want to avoid.